### ISSUE TITLE
Crash when moving an entity with many physics shapes

### ISSUE THREAD

**Issue by ElementalAndy** [OurMachinery/themachinery-public#1272]
- **Version:** Git Master Branch
- **OS:** Windows

I get a race condition with concurrent writes to physx when moving a fairly complex entity around in the scene view.

One thread comes from gizmo_set_transform(..) in transform_component.c and the other one from private__mirror_static_transforms_engine(..)  in physx_scene.cpp.

![gizmo_callstack](https://user-images.githubusercontent.com/91697782/163359314-02a98279-c4f9-40e2-aba9-1b5cddb46274.png)
![mirror_static_transforms_callstack](https://user-images.githubusercontent.com/91697782/163359364-a1ee20eb-ff1a-4a38-beb6-187ac6e7a3e8.png)

**Comment by niklas-ourmachinery** [comment]
Investigating this, it looks like `tm_entity_api->update()` does not wait for all engine/system jobs to complete before it exits. That means that the rest of the Scene Tab code runs in parallel with the entity system update. We get a race condition in physics when the Scene Tab tries to move a physics object while the physics update job is still running.

To verify that this is the problem, you can:

1. Add `tm_os_api->thread->sleep(0.1);` to the start of `update()` in `physx_scene.cpp`.
2. Put a breakpoint on the line `if (sim->mixer)` after the `tm_entity_api->update()` call in `simulation.c`.

You will notice that when the breakpoint is hit, one of the job threads is still at the `sleep()` in `physx_scene.cpp`, i.e. the `update()` call has exited without waiting for all the jobs/systems to finish.

This issue can probably cause lots of other problematic race conditions too.

**Comment by niklas-ourmachinery** [comment]
I'm a bit confused, because in `private__update_stage()` it looks like it waits for all jobs to complete before continuing:

~~~c
    for (uint32_t si = 0; si < n; ++si) {
#if defined(TM_NO_MAIN_FIBER)
        ctx->job_system_api->wait_for_counter_and_free_from_os_thread(counters[si], 0.0);
#else
        ctx->job_system_api->wait_for_counter_and_free(counters[si]);
#endif
        atomic_fetch_add_uint32_t(&ctx->world_tick, 1);
    }
~~~

@simon-ourmachinery or @tobias-ourmachinery ... any ideas?

**Comment by simon-ourmachinery** [comment]
> Just a note on this, we are not using `simulation.c` in our runtime environment, we have a very simplified copy paste of it, and it was copied half a year ago so it starts to grow old. Any synchronization related things which simulation.c deals with, and bugs which get fixed there, these are things which everyone implementing their own engine on top of you plugins needs to know about so they can make the same fixes in their version of simulation.c
> 
> Just pointing this out to emphasise that documentation about any explicit synchronization needs, i.e it is not automatically handled by the ECS framework, that is very important to know about.

The ECS framework should be self contain when it comes to sync. All functions state by now when to use them in a thread context.



> I'm a bit confused, because in `private__update_stage()` it looks like it waits for all jobs to complete before continuing:
> 
> ```c
>     for (uint32_t si = 0; si < n; ++si) {
> #if defined(TM_NO_MAIN_FIBER)
>         ctx->job_system_api->wait_for_counter_and_free_from_os_thread(counters[si], 0.0);
> #else
>         ctx->job_system_api->wait_for_counter_and_free(counters[si]);
> #endif
>         atomic_fetch_add_uint32_t(&ctx->world_tick, 1);
>     }
> ```
> 
> @simon-ourmachinery or @tobias-ourmachinery ... any ideas?

Stages are supposed to wait (as the code indicates) till all engines/systems are done. A possible problem is that we are using a notify call. They are not thread save unless NO engine/system is running. They are used for example in the transform API (`update_local_transform()`). In both cases, it is not recommended from looking at the ECS code to call them from the outside of the ECS when a system/engine might run. There are other APIs that suffer from the same issue. Which we are aware of and in our current review of the ECS we have discussed addressing this. A fix for this would be (which might cause issues) to cache all calls to notify and execute them either at the end of the current stage. This would make sure that they cannot interfere with any of the systems/engines scheduled in parallel at the moment. 

It might cause some systems not to get the right data that rely on the notify system since the data might be modified by the end of the stage by a different engine/system in that frame.

These reasons were some of the motivations for implementing the "reactive" system in which engines run on changes or adding or removal of components.

**Comment by Tisten** [comment]
This issue has been fixed by my PR https://github.com/OurMachinery/themachinery/pull/976, and here https://github.com/OurMachinery/themachinery/pull/976/files#r873551040 I wrote a comment about it fixing https://github.com/OurMachinery/themachinery-public/issues/1312 which is a duplicate of this one.

So please help me close this (I'm not allowed to)
